File: sol/solution/0001-0100/0040. Combination Sum II/Solution.kt
class Solution {
    fun combinationSum2(candidates: IntArray, target: Int): List<List<Int>> {
        val result = mutableListOf<List<Int>>()
        candidates.sort()
        backtrack(candidates, target, 0, mutableListOf(), result)
        return result
    }
    private fun backtrack(
        candidates: IntArray,
        target: Int,
        start: Int,
        current: MutableList<Int>,
        result: MutableList<List<Int>>
    ) {
        if (target == 0) {
            result.add(ArrayList(current))
            return
        }
        for (i in start until candidates.size) {
            if (i > start && candidates[i] == candidates[i - 1]) continue
            if (candidates[i] > target) break
            current.add(candidates[i])
            backtrack(candidates, target - candidates[i], i + 1, current, result)
            current.removeAt(current.size - 1)
        }
    }
}

File: sol/solution/0001-0100/0040. Combination Sum II/Solution.rs
impl Solution {
    pub fn combination_sum2(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut candidates = candidates;
        candidates.sort();
        let mut result = Vec::new();
        let mut current = Vec::new();
        Self::backtrack(&candidates, target, 0, &mut current, &mut result);
        result
    }
    fn backtrack(candidates: &[i32], target: i32, start: usize, current: &mut Vec<i32>, result: &mut Vec<Vec<i32>>) {
        if target == 0 {
            result.push(current.clone());
            return;
        }
        for i in start..candidates.len() {
            if candidates[i] > target {
                break;
            }
            if i > start && candidates[i] == candidates[i - 1] {
                continue;
            }
            current.push(candidates[i]);
            Self::backtrack(candidates, target - candidates[i], i + 1, current, result);
            current.pop();
        }
    }
}